concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -161,37 +159,46 @@ processFinalization newFinBlock finRec@FinalizationRecord{..} = do
-                toRemove2 <- pruneTrunk toRemove1 parent brs
-                -- mark blocks as finalized now, so that blocks are marked finalized by increasing height
-                markFinalized (getHash keeper) finRec
-                logEvent Skov LLDebug $ "Block " ++ show keeper ++ " marked finalized"
-                -- Finalize the transactions of the surviving block.
-                -- (This is handled in order of finalization.)
-                finalizeTransactions (getHash keeper) (blockSlot keeper) (blockTransactions keeper)
-                ati <- bpTransactionAffectSummaries keeper
-                bcTime <- getSlotTimestamp (blockSlot keeper)
-                let ctx = BlockContext{
-                        bcHash = getHash keeper,
-                        bcHeight = bpHeight keeper,
-                        ..}
-                flushBlockSummaries ctx ati =<< getSpecialOutcomes =<< blockState keeper
-                return toRemove2
-
-        toRemoveFromTrunk <- pruneTrunk [] newFinBlock (Seq.take pruneHeight oldBranches)
+                       -> m ([BlockPointerType m], [BlockPointerType m])
+                       -- ^ The return value is a list of blocks to mark dead (ordered by decreasing
+                       -- height) and a list of blocks to mark finalized (ordered by increasing
+                       -- height).
+            pruneTrunk toRemove toFinalize _ Seq.Empty = return (toRemove, toFinalize)
+            pruneTrunk toRemove toFinalize keeper (brs Seq.:|> l) = do
+              let toRemove1 = toRemove ++ filter (/= keeper) l
+              parent <- bpParent keeper
+              pruneTrunk toRemove1 (keeper : toFinalize) parent brs
+        (toRemoveFromTrunk, toFinalize) <- pruneTrunk [] [] newFinBlock (Seq.take pruneHeight oldBranches)
+        -- Add the finalization to the finalization list
+        addFinalization newFinBlock finRec
+        -- mark blocks as finalized in the order returned by `pruneTrunk`, so that blocks are marked
+        -- finalized by increasing height
+        mfs <- forM toFinalize $ \block -> do
+          markFinalized (getHash block) finRec
+        -- Finalize the transactions of surviving blocks in the order of their finalization.
+        fts <- forM toFinalize $ \block -> do
+          finalizeTransactions (getHash block) (blockSlot block) (blockTransactions block)
+        forM_ toFinalize $ \block -> do
+          ati <- bpTransactionAffectSummaries block
+          bcTime <- getSlotTimestamp (blockSlot block)
+          let ctx = BlockContext{
+                bcHash = getHash block,
+                bcHeight = bpHeight block,
+                ..}
+          flushBlockSummaries ctx ati =<< getSpecialOutcomes =<< blockState block
+        -- block states and transaction statuses need to be added into the same LMDB transaction
+        -- with the finalization record, if persistent tree state is used
+        wrapupFinalization mfs fts
```

Comment:
Is there a reason this function does not take a finalization record and instead looks it up?

I believe passing it here explicitly would be conceptually cleaner and harder to screw up.

Reply:
At one point, I was thinking to pass lists of blocks and transactions to be finalized through the tree state, just like the last finalization record already was. Then I figured that it's probably not optimal because these lists would only be used by `wrapupFinalization`. But since the last finalization record was part of the (persistent) tree state anyway, I could just look it up.

Since `wrapupFinalization` is only doing anything in its persistent implementation, I also wanted to make its signature "quieter" (having fewer parameters). But of course, passing a parameter explicitly makes the data flow more obvious than passing it through the state. If you think it's better that way, I'll make it explicit.

Reply:
I would prefer it, but it is your decision.